[consumer] Allow annotating consumer errors with metadata (#9041)
open-telemetry/opentelemetry-collector: consumer/consumererror/README.md

the error for use upstream in the pipeline. Ideally the error returned by a
component in its `consume` function should be from this package.

## Use cases

**Retry logic**: Errors should be allowed to embed rejected telemetry items to
be retried by the consumer.

**Indicating partial success**: Errors can indicate that not all items were
accepted, for example as in an OTLP partial success message.

**Communicating network error codes**: Errors should allow embedding information
necessary for the Collector to act as a proxy for a backend, i.e. relay a status
code returned from a backend in a response to a system upstream from the
Collector.

## Creating Errors

Errors can be created through one of the three constructors provided:

- `consumererror.New[Signal]` for retryable errors.
- `consumererror.NewPartial` for errors where only some of the items failed.
- `consumererror.NewHTTPStatus` for errors resulting from an HTTP call with an error status code.
- `consumererror.NewGRPCStatus` for errors resulting from a gRPC call with an error status code.

Only retryable errors will be retried, all other errors are considered permanent

> and will not be retried.

Reviewer: Do we need to make the new error types have some sort of relationship with `IsPermanent`? Right now if you passed a `consumererror.NewHTTPStatus` to `consumererror.IsPermanent` it will return false.

Author: I was expecting that we would get rid of `IsPermanent` since we can just assume an error is permanent unless we are retrieving the retryable signal data from the error. We should handle these in the meantime, I've updated the function for now to transition us.

component in its `consume` function should be from this package.

## Use cases

**Retry logic**: Errors should be allowed to embed rejected telemetry items to
be retried by the consumer.

**Indicating partial success**: Errors can indicate that not all items were
accepted, for example as in an OTLP partial success message.

**Communicating network error codes**: Errors should allow embedding information
necessary for the Collector to act as a proxy for a backend, i.e. relay a status
code returned from a backend in a response to a system upstream from the
Collector.

## Creating Errors

Errors can be created through one of the three constructors provided:

- `consumererror.New[Signal]` for retryable errors.
- `consumererror.NewPartial` for errors where only some of the items failed.
- `consumererror.NewHTTPStatus` for errors resulting from an HTTP call with an error status code.
- `consumererror.NewGRPCStatus` for errors resulting from a gRPC call with an error status code.

Only retryable errors will be retried, all other errors are considered permanent
and will not be retried.

Errors can be joined by passing them to a call to `errors.Join`.

> ### Retryable errors

Author: I realized we haven't documented what to do if there is a discrepancy between the data held by the caller and the consumer. The only behavior that I think will be consistent would be to only consider the data returned by the consumer. Otherwise we should remove this parameter from the `New[Signal]` functions and just use the copy held by the caller. If anyone has other thoughts on this, please let me know.

Author: I think these two approaches would be our best bet:

1. Continue to require data in `New[Signal]` and indicate to caller implementors that they can discard their copy of the data and rely on the consumer to return a copy. This requires consumers to keep track of the data they submitted.
2. Add a `WithRetryData` option or similar that indicates exactly which data to retry, allowing consumers to specify a subset of data if they want, and remove signal data as a required parameter in the constructor. Callers are expected to keep a copy of the data submitted to each consumer, but if they receive this option are expected to use the consumer's copy instead. I think this is more in line with how things currently work.

Author: Per an offline discussion on 2024-04-17, it was agreed to take approach 1.

# Consumer errors

This package contains error types that should be returned by a consumer when an
error occurs while processing telemetry. The error types included in this
package provide functionality for communicating details about
the error for use upstream in the pipeline. Ideally the error returned by a
component in its `consume` function should be from this package.

## Use cases

**Retry logic**: Errors should be allowed to embed rejected telemetry items to
be retried by the consumer.

**Indicating partial success**: Errors can indicate that not all items were
accepted, for example as in an OTLP partial success message.

**Communicating network error codes**: Errors should allow embedding information
necessary for the Collector to act as a proxy for a backend, i.e. relay a status
code returned from a backend in a response to a system upstream from the
Collector.

## Creating Errors

Errors can be created through one of the three constructors provided:

- `consumererror.New[Signal]` for retryable errors.
- `consumererror.NewPartial` for errors where only some of the items failed.

> - `consumererror.NewHTTPStatus` for errors resulting from an HTTP call with an error status code.
> - `consumererror.NewGRPCStatus` for errors resulting from a gRPC call with an error status code.

Reviewer 2: ```suggestion
- `consumererror.NewHTTPError` for errors resulting from an HTTP call with an error status code.
- `consumererror.NewGRPCError` for errors resulting from a gRPC call with an error status code.
```

Same in other places.

Should we make it even shorter `consumererror.New[GRPC|HTTP]`?

Author: My bad, the error names already ended in `Error`. Either way, I like the shorter names: they don't stutter and match the other error names.

Reviewer 2: Sounds good!

Only retryable errors will be retried, all other errors are considered permanent
and will not be retried.

Errors can be joined by passing them to a call to `errors.Join`.

### Per-signal retryable errors

If an error is considered transient and data processing should be retried, the data
that should be retried should be created using `NewTraces`, `NewMetrics` or `NewLogs`
with the data that should be retried, e.g. `NewTraces(err, traces)`. The data that
is contained in the error will be the only data that is retried, the caller should
not use its copy of the data for the retry.

## Other considerations

To keep error analysis simple when looking at an error upstream in a pipeline,
the component closest to the source of an error or set of errors should make a
decision about the nature of the error. The following are a few places where
special considerations may need to be made.

### Fanouts

#### Permanent errors

When a fanout component receives multiple permanent errors from downstream in
the pipeline, it should determine the error with the highest precedence and

> forward that upward in the pipeline. If a precedence cannot be determined,

Reviewer 2: What's the error precedence? I don't think we define it anywhere

Author: It's up to the component to determine which error is most important. The intent of this statement is that If I'm running a Collector as a proxy where data is forwarded synchronously, I can only return a single error code even if I have multiple downstream endpoints. So if I get a 404 and two 400s from three downstream endpoints, then some component in the pipeline needs to determine whether to return a 404, a 400, or some third code.

the component should forward all errors.

#### Retryable errors

When a fanout component receives a retryable error, it must only retry the data

> the failed pipeline branch.

Reviewer 2: Fanout doesn't retry itself. So it must include some information about the error destination if we want to achieve this

Author: The fanout consumer doesn't retry itself, but I think it has the necessary context to make the retries since it knows which component it sends to in its [consume function](https://github.com/open-telemetry/opentelemetry-collector/blob/main/internal/fanoutconsumer/traces.go#L47). Otherwise we can do retries on individual pipeline branches like in the [exporter helper](https://github.com/open-telemetry/opentelemetry-collector/blob/main/exporter/exporterhelper/retry_sender.go#L95).

Do you think we may miss anything here that would necessitate including information about the failed consumer in the error?

Reviewer 2: The fanout doesn't have a state, it just responds with an error (or nil) upstream once all the downstream consumers have responded. Passing information (as part of the joined error) about what consumer failed is one problem, another problem is to use that information by a component that is going to do the retry and pass it back to the fanout (likely with context) so the fanout can retry only the failed destination.

There is nothing to do to address it in this PR. My idea is that we need to think about the potential expansion of the error type to add that information in the future.

Author: I added `consumererror.Component` to communicate this upstream in the pipeline.

#### Mixed errors

When a mix of retryable and permanent errors are received from downstream
pipeline branches, the fanout component should continue all retries for
retryable data and return a permanent error upstream regardless of whether
any of the retries succeed. The component performing retries should capture
which pipelines succeeded or failed using the Collector's pipeline observability
mechanisms.

### Signal conversion

When converting between signals in a pipeline, it is expected that the component
performing the conversion should perform the translation necessary in the error
for any signal item counts. If the converted count cannot be determined, the full
count of pre-converted signals should be returned.

### Asynchronous processing

The use of any components that do asynchronous processing in a pipeline will cut
off error backpropagation at the asynchronous component. The asynchronous
component may communicate error information using the Collector's own signals.

## Examples

Creating an error:

```golang
errors.Join(
    consumererror.NewTraces(
        error,
        traces,
        consumererror.WithRetryDelay(2 * time.Duration)
    ),

>     consumererror.NewHTTPStatus(error, http.StatusTooManyRequests)

Reviewer 2: ```suggestion
    consumererror.NewHTTPError(error, http.StatusTooManyRequests)
```

)
```

Using an error:

```golang
err := nextConsumer(ctx, traces)

> var status consumererror.StatusError

Reviewer 3: ```suggestion
var status consumererror.NetworkError
```

component in its `consume` function should be from this package.

## Use cases

**Retry logic**: Errors should be allowed to embed rejected telemetry items to
be retried by the consumer.

**Indicating partial success**: Errors can indicate that not all items were
accepted, for example as in an OTLP partial success message.

**Communicating network error codes**: Errors should allow embedding information
necessary for the Collector to act as a proxy for a backend, i.e. relay a status
code returned from a backend in a response to a system upstream from the
Collector.

## Creating Errors

Errors can be created through one of the three constructors provided:

- `consumererror.New[Signal]` for retryable errors.
- `consumererror.NewPartial` for errors where only some of the items failed.
- `consumererror.NewHTTP` for errors resulting from an HTTP call with an error status code.
- `consumererror.NewGRPC` for errors resulting from a gRPC call with an error status code.
- `consumererror.NewComponent` for metadata around the component that produced the error.

Only retryable errors will be retried, all other errors are considered permanent

> and will not be retried.

Reviewer 2: We cannot assume this behavior. Probably "should" or "must" would be better fit here?
```suggestion
and should not be retried.
```

pipeline branches, the fanout component should continue all retries for
retryable data and return a permanent error upstream regardless of whether
any of the retries succeed. The component performing retries should capture
which pipelines succeeded or failed using the Collector's pipeline observability
mechanisms.

### Signal conversion

When converting between signals in a pipeline, it is expected that the component
performing the conversion should perform the translation necessary in the error
for any signal item counts. If the converted count cannot be determined, the full
count of pre-converted signals should be returned.

### Asynchronous processing

The use of any components that do asynchronous processing in a pipeline will cut
off error backpropagation at the asynchronous component. The asynchronous
component may communicate error information using the Collector's own signals.

## Examples

Creating an error:

```golang
errors.Join(
    consumererror.NewComponent(exp.componentID),
    consumererror.NewTraces(
        error,
        traces,
        consumererror.WithRetryDelay(2 * time.Duration)
    ),

>     consumererror.NewHTTP(error, http.StatusTooManyRequests)

Reviewer 2: I don't think such composition is the right approach to communicating a particular error. How is a component handling an error supposed to understand if this is one error or three errors from a fanout? I think we should keep `errors.Join` only for fanout and not to compose a particular error. 

Can we have the component, retryable/not-retryable, grps/http be fields of one error type?

Author: >Can we have the component, retryable/not-retryable, grps/http be fields of one error type?

This was the original approach, but I modified it in response to https://github.com/open-telemetry/opentelemetry-collector/pull/9041#discussion_r1417768965. I still agree with my initial assessment that it still feels a bit odd, though it does perhaps provide some flexibility.

>How is a component handling an error supposed to understand if this is one error or three errors from a fanout?

I still think the fanout is the only thing that can meaningfully handle this since it is the only place where we have the context of where the error came from. Even if we put all fields on a single error object, the component will still only get an list of errors with unstable ordering and will have to e.g. resend data to call components if there's a retryable error. In my opinion, if there's a fanout, we need to have the fanout component make the determination of which error to forward upstream.

Author: If the fanout can't determine a precedence, you can still get each individually wrapped error using the `Unwrap` method on one of the wrapped errors returned by the fanout, though maybe this isn't a strong enough guarantee.

an error is not considered permanent.

#### WithRetryDelay(delay time.Duration)

Indicate that the payload should be retried after a certain amount of time.

### WithHTTPStatus(int)

Annotate the error with an HTTP status code obtained from a response during
exporting.

### WithGRPCStatus(*status.Status)

Annotate the error with a gRPC status code obtained from a response during
exporting.

## Reading from an error

The `consumererror.Error` type supports the following methods. Each method has a
method signature like `(data, bool)` with the second value indicating whether
the option was passed during the error's creation.

- `Count`: Get the count of rejected items.
- `Retryable[Signal]`: Gets the information necessary to retry the request for a
  given signal.
- `ToHTTP`: Returns an integer representing the HTTP status code. If both
  `WithHTTPStatus` and `WithGRPCStatus` were passed, the HTTP status is used
  first and gRPC status second.
- `ToGRPC`: Returns a `*status.Status` object representing the status returned
  from the gRPC server. If both `WithHTTPStatus` and `WithGRPCStatus` were
  passed, the gRPC status is used first and HTTP status second.

## Other considerations

> ### Asynchronous processing

Reviewer 4: Are we able to account for non-linear data flow in some way? For example, when a receiver is shared by multiple pipelines, an error may be returned from any of these pipelines, but I believe we currently do not have a generalized ability to retry only to the pipeline which returned the error. We may need such a mechanism within the consumer package (or fanoutconsumer) in order to avoid sending duplicate data in response to errors.

Author: There are two fanout locations that may want to utilize these errors to do retries:
* [fanoutconsumer](https://github.com/open-telemetry/opentelemetry-collector/blob/edf7aca8522a8396ec155eabf8c1119690ed6e2d/internal/fanoutconsumer/traces.go#L47)
* [exporterhelper](https://github.com/open-telemetry/opentelemetry-collector/blob/cc88aee67530478255ecd4f4f5930390c88f5cb1/exporter/exporterhelper/retry_sender.go#L25)

I've tried my best to document the possible situations we may encounter and advice on how to handle them. If I've missed or misrepresented anything, let me know.

accepted, for example as in an OTLP partial success message.

**Communicating network error codes**: Errors should allow embedding information
necessary for the Collector to act as a proxy for a backend, i.e. relay a status
code returned from a backend in a response to a system upstream from the
Collector.

## Current targets for using errors

**Receivers**: Receivers should be able to consume multiple errors downstream
and determine the best course of action based on the user's configuration. This
may entail either keeping the retry queue inside of the Collector by having the
receiver keep track of retries, or may involve having the caller manage the
retry queue by returning a retryable network code with relevant information.

**exporterhelper**: When an exporter returns a retryable error, the
exporterhelper can use this information to manage the sending queue if it is
enabled. Permanent errors will be forwarded back up the pipeline.

## Creating Errors

Errors can be created by calling `consumererror.New(err, opts...)` where `err`
is the underlying error, and `opts` is one of the provided options for supplying
additional metadata:

- `consumererror.WithRetry`

> - `consumererror.WithPartial`

Author: I'll add a section about this. I think we have to assume that the user knows what they are doing, even if we likely consider it an error. The rule is that an error is retryable if it includes retry information, so in this case we have to assume that a component had partial success, but also supports retries on partial success, even if OTLP considers partial success permanent.

> - `consumererror.WithPartial`

Reviewer 5: Let's assume is retryable, what data are you retrying? Based on my understanding "partial" only contains the number of failed entries, so you cannot retry because you don't know what failed.

> - `consumererror.WithPartial`

Author: You would have to retry the whole payload, and the backend would need to support filtering out the items that have already been accepted. From the perspective of a component upstream from where this error was created, we don't know anything about where the data will be sent and will just have to work with the available context.

I don't expect anyone to actually do this, but I also don't see any meaningful way for us to prevent it. The best we could likely do would be to swallow one of the options and log a warning on the console if we can determine a precedence for the errors, but I would still prefer to leave that to upstream components to deal with. I'd also prefer to keep things resilient and not return an error while creating the error.

Even if we break out the error cases into separate types, if you have to join anything with `errors.Join`, you're back in the same place where contradictory errors can be placed next to each other.

- `consumererror.WithGRPCStatus`

> - `consumererror.WithHTTPStatus`

Author: If a component uses both of these, the assumption is that it has its own HTTP<->gRPC translation, and wants to specify both statuses. The error will produce the given status if it has it, otherwise will convert from the status from the other transport.

- `consumererror.WithMetadata`

**Example**:

```go
consumererror.New(err,
  consumererror.WithRetry(
    componentID,
    consumerrerror.WithRetryDelay(10 * time.Second)
  ),
  consumererror.WithGRPCStatus(codes.InvalidArgument),
)
```

### Retrying data submission

*NOTE*: This section is a draft and will be developed separately from the rest of this proposal.

If an error is transient, the `WithRetry` option corresponding to the relevant
signal should be used to indicate that the error is retryable and to pass on any
retry settings. These settings can come from the data sink or be determined by
the component, such as through runtime conditions or from user settings.

The data for the retry will be provided by the component performing the retry.

**Note**: If retry information is not included in an error, the error will be
considered permanent and will not be retried.

**Usage:**

```go
consumererror.WithRetry(

>   componentID,

Reviewer 5: What do I do with the "component" if I know that a specific component failed? Especially that I don't know the structure of the pipelines.

Author: That's actually a good point, we should have some other pipeline component add the component ID automatically. The goal is that we will be including the failed pipeline branch inside this error so only retryable branches are retried.

Note that retryable errors are outside the scope of what will be implemented in this PR, they're a lot more complicated than e.g. HTTP/gRPC codes.

component reads a code, it will read a code for its desired transport, and the
code may be translated depending whether the input and output transports are
different. For example, a gRPC exporter may record a gRPC status. If a gRPC
receiver reads this status, it will be exactly the provided status. If an HTTP
receiver reads the status, it wants an HTTP status, and the gRPC status will be
converted to an equivalent HTTP code.

**Usage:**

```go
consumererror.WithGRPCStatus(codes.InvalidArgument)
consumererror.WithHTTPStatus(http.StatusTooManyRequests)
```

### Including custom data

Custom data can be included as well for any additional information that needs to
be propagated back up the pipeline. It is up to the consuming component if or
how this data will be used.

**Usage:**

```go
consumererror.WithMetadata(MyMetadataStuct{})
```

To keep error analysis simple when looking at an error upstream in a pipeline,
the component closest to the source of an error or set of errors should make a
decision about the nature of the error. The following are a few places where
special considerations may need to be made.

## Using errors

### Fanouts

> When fanouts receive multiple errors, they will combine them with

Author: "receive" is right here, but I can make "fanout" singular. I'll update this.

# Consumer errors

This package contains error types that should be returned by a consumer when an
error occurs while processing telemetry. The error types included in this
package provide functionality for communicating details about
the error for use upstream in the pipeline. Ideally the error returned by a
component in its `consume` function should be from this package.

## Error classes

**Retryable**: Errors are retryable if re-submitting data to a sink may result

> in a successful submission.

Reviewer 5: This needs to carry some informations:
* [required] what to retry on;
* [optional] retry interval?

Author: This is just a classification of error, I cover the details of what retryable errors contain here: https://github.com/open-telemetry/opentelemetry-collector/pull/9041/files#diff-e0fa8222784a2c0c2683b70e2b6a7ccf2c54b6477acd7e2518839e38060bdcf5R90.

We discussed this last week and determined that the caller can provide the data to retry, but we're going to focus on exactly what goes into a retryable error after this PR. Right now we're mainly focusing on the HTTP/gRPC errors.

**Permanent**: Errors are permanent if submission will always fail for the
current data. Errors are considered permanent unless they are explicitly marked

> as retryable.

Reviewer 5: For permanent errors, we should probably always return the "number of not accepted" entries? Is that reasonable?

Author: I don't think we need to do that since the caller will already know how many items failed.

## Use cases

**Retry logic**: Errors should be allowed to include information necessary to
perform retries.

**Indicating partial success**: Errors can indicate that not all items were

> accepted, for example as in an OTLP partial success message.

Reviewer 5: Add clarification that, in order to implement the OTLP protocol we need to carry this information to the receiver.

Author: Done.

**Communicating network error codes**: Errors should allow embedding information
necessary for the Collector to act as a proxy for a backend, i.e. relay a status
code returned from a backend in a response to a system upstream from the

> Collector.

Reviewer 5: I think @jpkrohling had a more detailed use-case here, but would be interesting to hear what problem we are solving. If I remember correctly was something that the collector should not retry and all retries to be done by the caller, but cannot remember exactly, so better to document in details here.

Author: I can provide more examples if you like, but the simplest one is the one I described here: the Collector can act as a proxy and relay a code from a backend back to the caller. This works for retries too, if the code is e.g. an HTTP `429` status code.

> ## Current targets for using errors

Reviewer 5: Scraper helper needs this for scraping errors as well.

Author: Done.

**Receivers**: Receivers should be able to consume multiple errors downstream
and determine the best course of action based on the user's configuration. This
may entail either keeping the retry queue inside of the Collector by having the
receiver keep track of retries, or may involve having the caller manage the
retry queue by returning a retryable network code with relevant information.

**exporterhelper**: When an exporter returns a retryable error, the
exporterhelper can use this information to manage the sending queue if it is
enabled. Permanent errors will be forwarded back up the pipeline.

**obsreport**: Recording partial success information can ensure we correctly
track the number of failed telemetry records in the pipeline. Right now, all
records will be considered to be failed, which isn't accurate when partial
successes occur.

## Creating Errors

Errors can be created by calling `consumererror.New(err, opts...)` where `err`
is the underlying error, and `opts` is one of the provided options for supplying
additional metadata:

- `consumererror.WithRetry`
- `consumererror.WithPartial`
- `consumererror.WithGRPCStatus`
- `consumererror.WithHTTPStatus`

> - `consumererror.WithMetadata`

Reviewer 5: I am not seeing any use-case for this, why do we need it?

Author: This is for extensibility. I don't think we would add this until it was requested, but basically a custom exporter and receiver could communicate using a custom struct type by putting the struct inside the error with `WithMetadata`.

Reviewer 2: Maybe remove it for now? We should problably ensure the model is extendable but not over-prescriptive for something we don't have yet

Author: If extensibility is an explicit goal, I'd like to keep this in the design doc. I've added warnings to things that don't actually exist right now.

All options can be combined, we assume that the component knows what it is doing
when seemingly conflicting options.

Two examples:

- `WithRetry` and `WithPartial` are included together: Partial successes are
  considered permanent errors in OTLP, which conflicts with making an error
  retryable by including `WithRetry`. However, per our definition of what makes
  a permanent error, this error has been marked as retryable, and therefore we
  assume the component producing this error supports retyable partial success

>   errors.

Reviewer 5: See my previous question about this and what it means if I know that 5 out of 10 spans failed to send.

Author: If I know that 5/10 spans failed, then I can record that in `obsreport`, but will resubmit all 10; maybe the backend uses the span IDs to make span submission idempotent. I don't have a real use-case in mind for this, what I want to cover in this section is that we can use the rules established here to handle seemingly invalid error combinations.

Reviewer 5: > I don't have a real use-case in mind for this, what I want to cover in this section is that we can use the rules established here to handle seemingly invalid error combinations.

When in doubt leave it out. I want us to not add any API if we don't have a good use-case.

Author: Agreed. This doesn't add any API, the point of this paragraph is just to explain how seemingly conflicting error combinations could be reconciled.

can be propagated up to a receiver and returned to the caller.

**Communicating network error codes**: Errors should allow embedding information
necessary for the Collector to act as a proxy for a backend, i.e. relay a status
code returned from a backend in a response to a system upstream from the
Collector.

## Current targets for using errors

**Receivers**: Receivers should be able to consume multiple errors downstream
and determine the best course of action based on the user's configuration. This
may entail either keeping the retry queue inside of the Collector by having the
receiver keep track of retries, or may involve having the caller manage the
retry queue by returning a retryable network code with relevant information.

**scraperhelper**: The scraper helper can use information about errors from
downstream to affect scraping. For example, if the backend is struggling with
the volume of data, scraping could be slowed, or the amount of data collected
could be reduced. 

**exporterhelper**: When an exporter returns a retryable error, the
exporterhelper can use this information to manage the sending queue if it is

> enabled. Permanent errors will be forwarded back up the pipeline.

Author: My intention with this comment was that we would use these errors instead of the `throttleRetry` type, but I'm fine not calling out the sending queue specifically here.

`(consumererror.Error).Combine(errs...)` and pass them upstream. The upstream
component can then later pull all errors out for analysis.

### Retrieving errors

When a receiver gets a response that includes an error, it can get the data out
by doing something similar to the following. Note that this uses the `ErrorData`
type, which is for reading error data, as opposed to the `Error` type, which is
for recording errors.

```go
cerr := consumererror.Error{}
var errData []consumerError.ErrorData 

if errors.As(err, &cerr) {
  errData := cerr.Data()

  for _, data := range errData {
    data.HTTPStatus()
    data.Retryable()
    data.Partial()
  }
}
```

### Error data

Obtaining data from an error can be done using an interface that looks something like this:

```go
type ErrorData interface {
  // Returns the underlying error
  Error() error

  // Returns nil if a status is not available.
  HTTPStatus() *int

  // Returns nil if a status is not available.
  GRPCStatus() *status.Status

  // Returns nil if the error contains no retry information.
  Retryable() *Retryable

  // Returns a count of failed records or nil if no partial success information was recorded. 

>   Partial() *int

Reviewer 2: Shouldn't we have a `Partial` struct similar to `Retryable` so we can extend it if needed?

Author: Good with me, done.

when seemingly conflicting options.

Two examples:

- `WithRetry` and `WithPartial` are included together: Partial successes are
  considered permanent errors in OTLP, which conflicts with making an error
  retryable by including `WithRetry`. However, per our definition of what makes
  a permanent error, this error has been marked as retryable, and therefore we
  assume the component producing this error supports retyable partial success
  errors.
- `WithGRPCStatus` and `WithHTTPStatus` are included together: While the
  component likely only sent data over one of these transports, our errors will
  produce the given status if it is included on the error, otherwise it will
  translate a status from the status for the other transport. If both of these
  are given, we assume the component wanted to perform its own status
  conversion, and we will simply give the status for the requested transport
  without performing any conversion.

**Example**:

```go
consumererror.New(err,
  consumererror.WithRetry(
    consumerrerror.WithRetryDelay(10 * time.Second)
  ),
  consumererror.WithGRPCStatus(codes.InvalidArgument),

> )

Reviewer 2: This example must be invalid. `codes.InvalidArgument` should not be retryable. 

In general, it still seems like all these options provide too much flexibility, making it confusing to use and easy to misuse. 

For example, why GRPC and HTTP have to be options? Why can't we just have different constructors like `consumererror.NewFromGRPC(grpc.Status)`? The status should have the retry information that we can retrieve if applicable. If we expect exporters to add extra info on top of grpc error, they can manually parse the grpc status and create a collector "internal" error another way.
